### ISSUE TITLE
partial fc load pre-trained model to continue train

### ISSUE THREAD

**Issue by fucker007** [deepinsight/insightface#1354]
I have sovled this problem,

if anyone wanna load checkpoint file,and continue train, maybe  this information will help you.

-  **`step 1`**  change config.sh like this:
```
export CUDA_VISIBLE_DEVICES='0,1,2,3,4,5,6,7'
export HOROVOD_GPU_ALLREDUCE=NCCL
export HOROVOD_GPU_ALLGATHER=NCCL
export HOROVOD_GPU_BROADCAST=NCLL
export MXNET_CPU_WORKER_NTHREADS=3

PYTHON_EXEC=/home/liuyang/anaconda3/bin/python
${PYTHON_EXEC} train_memory.py \
--dataset glint360k_8GPU \
--loss cosface \
--network r100 \
--models-root ./models/glint360k_8GPU_r100FC_1.0_fp16_cosface \
--pretrain True
```
-  **`step 2`**  chang ` train_memory.py` like this:
```
def parse_args():
    parser = argparse.ArgumentParser(description='Train parall face network')
    #general
    parser.add_argument('--dataset', default='emore', help='dataset config')
    parser.add_argument('--network', default='r100', help='network config')
    parser.add_argument('--loss', default='cosface', help='loss config')
    parser.add_argument('--pretrain', default='False', help='pretrain or not')

    args, rest = parser.parse_known_args()
    default.generate_config(args.loss, args.dataset, args.network)
    parser.add_argument('--models-root', default="./test", help='root directory to save model.')
    args = parser.parse_args()
    return args
def train_net():
     ........
    train_module = SampleDistributeModule(
        pretrain=args.pretrain,
        prefix = prefix_dir,
        symbol=esym,
        fc7_model=fc7_model,
        memory_bank=memory_bank,
        memory_optimizer=memory_bank_optimizer)

```
-  **`step 3`**  chang `memory_bank.py` like this
```
class MemoryBank(object):
    def __init__(self,pretrain,num_sample, num_local, rank, local_rank, embedding_size, prefix, gpu=True):
        self.num_sample = num_sample
        self.num_local = num_local
        self.rank = rank
        self.embedding_size = embedding_size
        self.gpu = gpu
        self.prefix = prefix

        if self.gpu:
            context = mx.gpu(local_rank)
        else:
            context = mx.cpu()

        # In order to apply update, weight and momentum should be storage.
        #print("num_local :",self.num_local)
        self.weight = nd.random_normal(
            loc=0, scale=0.01, shape=(self.num_local, self.embedding_size),
            ctx=context)
        self.weight_mom = nd.zeros_like(self.weight)
        #print("init weight type", self.weight)
        if pretrain == 'True':
            print("load center.... ", pretrain)
         self.load()
        self.weight_index_sampler = WeightIndexSampler(num_sample, num_local,  rank)
       def load(self):
        print("center param: ",os.path.join(self.prefix, "%d_centers.param" % self.rank))
        self.weight = nd.load(os.path.join(self.prefix, "%d_centers.param" % self.rank))[0]
        self.weight_mom = nd.load(os.path.join(self.prefix,"%d_centers_mom.param" % self.rank))[0]


```
- **`step 4`**
```
    def init_params(self, initializer, arg_params=None, aux_params=None,
                    allow_missing=False, force_init=False, allow_extra=False):
        if self.pretrain == 'True':
            _, arg_params, aux_params = mx.module.module.load_checkpoint(prefix=os.path.join(self.prefix,"model_average"),epoch=0)
            allow_missing = True
            allow_extra = True
        # backbone
        self.backbone_module.init_params(
            initializer=initializer, arg_params=arg_params,
            aux_params=aux_params, allow_missing=allow_missing,
            force_init=force_init, allow_extra=allow_extra)
```

_Originally posted by @fucker007 in https://github.com/deepinsight/insightface/issues/1283#issuecomment-740564556_

**Comment by anxiangsir** [comment]
Thank you very much for your contribution, maybe you can send a **pull request**?

**Comment by fucker007** [comment]
If you are flying lazy , just `git` code on this [https://github.com/fucker007/partial-fc-pretrian](url)

**Comment by fucker007** [comment]
> Thank you very much for your contribution, we will update the code according to your suggestions.

I have to say , I do'nt  update learning rate, I know learning rate will declay by step , so I handle this by my hand, If can auto load learning step information ,that is good.

**Comment by tranvanhoa533** [comment]
> I have sovled this problem,
> 
> if anyone wanna load checkpoint file,and continue train, maybe this information will help you.
> 
> * **`step 1`**  change config.sh like this:
> 
> ```
> export CUDA_VISIBLE_DEVICES='0,1,2,3,4,5,6,7'
> export HOROVOD_GPU_ALLREDUCE=NCCL
> export HOROVOD_GPU_ALLGATHER=NCCL
> export HOROVOD_GPU_BROADCAST=NCLL
> export MXNET_CPU_WORKER_NTHREADS=3
> 
> PYTHON_EXEC=/home/liuyang/anaconda3/bin/python
> ${PYTHON_EXEC} train_memory.py \
> --dataset glint360k_8GPU \
> --loss cosface \
> --network r100 \
> --models-root ./models/glint360k_8GPU_r100FC_1.0_fp16_cosface \
> --pretrain True
> ```
> 
> * **`step 2`**  chang ` train_memory.py` like this:
> 
> ```
> def parse_args():
>     parser = argparse.ArgumentParser(description='Train parall face network')
>     #general
>     parser.add_argument('--dataset', default='emore', help='dataset config')
>     parser.add_argument('--network', default='r100', help='network config')
>     parser.add_argument('--loss', default='cosface', help='loss config')
>     parser.add_argument('--pretrain', default='False', help='pretrain or not')
> 
>     args, rest = parser.parse_known_args()
>     default.generate_config(args.loss, args.dataset, args.network)
>     parser.add_argument('--models-root', default="./test", help='root directory to save model.')
>     args = parser.parse_args()
>     return args
> def train_net():
>      ........
>     train_module = SampleDistributeModule(
>         pretrain=args.pretrain,
>         prefix = prefix_dir,
>         symbol=esym,
>         fc7_model=fc7_model,
>         memory_bank=memory_bank,
>         memory_optimizer=memory_bank_optimizer)
> ```
> 
> * **`step 3`**  chang `memory_bank.py` like this
> 
> ```
> class MemoryBank(object):
>     def __init__(self,pretrain,num_sample, num_local, rank, local_rank, embedding_size, prefix, gpu=True):
>         self.num_sample = num_sample
>         self.num_local = num_local
>         self.rank = rank
>         self.embedding_size = embedding_size
>         self.gpu = gpu
>         self.prefix = prefix
> 
>         if self.gpu:
>             context = mx.gpu(local_rank)
>         else:
>             context = mx.cpu()
> 
>         # In order to apply update, weight and momentum should be storage.
>         #print("num_local :",self.num_local)
>         self.weight = nd.random_normal(
>             loc=0, scale=0.01, shape=(self.num_local, self.embedding_size),
>             ctx=context)
>         self.weight_mom = nd.zeros_like(self.weight)
>         #print("init weight type", self.weight)
>         if pretrain == 'True':
>             print("load center.... ", pretrain)
>          self.load()
>         self.weight_index_sampler = WeightIndexSampler(num_sample, num_local,  rank)
>        def load(self):
>         print("center param: ",os.path.join(self.prefix, "%d_centers.param" % self.rank))
>         self.weight = nd.load(os.path.join(self.prefix, "%d_centers.param" % self.rank))[0]
>         self.weight_mom = nd.load(os.path.join(self.prefix,"%d_centers_mom.param" % self.rank))[0]
> ```
> 
> * **`step 4`**
> 
> ```
>     def init_params(self, initializer, arg_params=None, aux_params=None,
>                     allow_missing=False, force_init=False, allow_extra=False):
>         if self.pretrain == 'True':
>             _, arg_params, aux_params = mx.module.module.load_checkpoint(prefix=os.path.join(self.prefix,"model_average"),epoch=0)
>             allow_missing = True
>             allow_extra = True
>         # backbone
>         self.backbone_module.init_params(
>             initializer=initializer, arg_params=arg_params,
>             aux_params=aux_params, allow_missing=allow_missing,
>             force_init=force_init, allow_extra=allow_extra)
> ```
> 
> _Originally posted by @fucker007 in [#1283 (comment)](https://github.com/deepinsight/insightface/issues/1283#issuecomment-740564556)_

Have you tried to resume training yet?

**Comment by jacqueline-weng** [comment]
Hi, I'm training with a 4-card machine and there is error when trying to load center.params 0-3. It is able to resume training if I only load model-000.params and model-symbol.json but still with some error messages like this:
`Read -1, expected 131072, errno = 1` multiple times.
Do you know how to adapt 8-card center.params to different number of cards? Any clues about these error messages?
